src/pudl/transform/ferc1.py

```diff
@@ -891,6 +883,68 @@ def reconcile_table_calculations(
     return calculated_df
 
 
+def calculate_values_from_components(
+    calculation_components: pd.DataFrame,
+    data: pd.DataFrame,
+    validate: Literal["one_to_many", "many_to_many"],
+    calc_idx: list[str],
+    data_idx: list[str],
+    value_col: str,
+) -> pd.DataFrame:
```

Comment:
name help here?!

Reply:
Feeling confused about why the additional dimensions aren't involved in merging the data and calculations.

```py
def apply_calculations(
    calculation_components: pd.DataFrame,
    data: pd.DataFrame,
    validate: Literal["one_to_many", "many_to_many"],
    calc_idx: list[str],
    data_idx: list[str],
    value_col: str,
) -> pd.DataFrame:
"""
Apply calculations derived from XBRL metadata to reported XBRL data.

Args:
    calculation_components: Table defining the calculations, with each row defining
        a single component, including its weight. Groups of rows identified by
        ``table_name_parent`` and ``xbrl_factoid_parent`` indicate the values being
        calculated.
    data: exploded FERC data to apply the calculations to. Primary key should be
        ``report_year``, ``utility_id_ferc1``, ``table_name``, ``xbrl_factoid``, and
        whatever additional dimensions are relevant to the data.
    validate: type of merge validation to apply when initially merging the calculation
        components (left) and the data (right). 
    calc_idx: primary key columns that uniquely identify a calculation component (not
        including the ``_parent`` columns).
    data_idx: primary key columns that uniquely identify a record in the data. (maybe
        this can be inferred automatically given calc_idx?)
    value_col: label of the column in ``data`` that contains the values to apply the
        calculations to (typically ``dollar_value`` or ``ending_balance``).
"""
...
```

Reply:
ty for the docs!!

Reply:
and the additional dimensions are involved by way of the `calc_idx`/`data_idx`